chainstate/src/detail/ban_score.rs

```diff
@@ -0,0 +1,143 @@
+use crate::BlockError;
+
+use super::{
+    pow::error::ConsensusPoWError, spend_cache::error::StateUpdateError, CheckBlockError,
+    CheckBlockTransactionsError, ConsensusVerificationError, OrphanCheckError,
+};
+
+pub trait BanScore {
+    fn ban_score(&self) -> u32;
+}
+
+impl BanScore for BlockError {
+    fn ban_score(&self) -> u32 {
+        match self {
+            BlockError::StorageError(_) => 0,
+            BlockError::OrphanCheckFailed(err) => err.ban_score(),
+            BlockError::CheckBlockFailed(err) => err.ban_score(),
+            BlockError::StateUpdateFailed(err) => err.ban_score(),
+            BlockError::BestBlockLoadError(_) => 0,
+            BlockError::InvariantErrorFailedToFindNewChainPath(_, _, _) => 0,
+            BlockError::InvariantErrorInvalidTip => 0,
+            BlockError::InvariantErrorPrevBlockNotFound => 0,
+            // Even though this should've been caught by orphans check, its mere presence means a peer sent a block they're not supposed to send
+            BlockError::PrevBlockNotFound => 100,
+            BlockError::InvalidBlockSource => 100,
```

Comment:
Is `100` the threshold at which a peer gets banned? Maybe there could be a constant for that.

Reply:
Yes. As described in the PR top message, the purpose for not having this be a constant is because the node operator will have the choice to change this to another value. It's a fair question what's the best way to write this number, but honestly I'm not too concerned about that point right now. It's just a constant being thrown around. I'm even considering making something like `thiserror` where we define the value in the error itself. Not the time for niceties right now though.

Reply:
Sure I see your points. I still think a named constant would be nice to establish some sort of reference value at which a peer gets banned by default, even if the numeric value is completely arbitrary

Reply:
> I still think a named constant would be nice

I'd probably vote for this over the magic number too in all honesty

Reply:
I'm fine with updating this later